### PR TITLE
Add bottom dialog pointer to color picker and fix positioning

### PR DESCRIPTION
**Pull request by volfied**
Using [CSS Mixin Guards](http://lesscss.org/features/#mixin-guards-feature), we now position the dialog pointer either at the top or the bottom of color picker dialog.

Also, addressing #3550, we now position the color picker correctly when there is no room below the clicked color input and color picker appears above it.

### CODE REVIEW

File: src/style/panel.less [hunk 1]
```diff
@@ -383,7 +383,7 @@
     width: 100%;
     
     &:before {
-        .dialog-pointer(5.4rem);
+        .dialog-pointer(5.4rem, true);
```
Comment (volfied) [inline src/style/panel.less]: I have to add this because there are no default parameters with less mixins.

File: src/style/shared/dialog.less [hunk 2]
```diff
@@ -190,10 +190,20 @@ dialog::backdrop {
     width: 100%;
 }
 
-.dialog-pointer(@left) {
+.pointerMixin (@atTop) when (@atTop) {
```
Comment (volfied) [inline src/style/shared/dialog.less]: This is equivalent to `@atTop === true` in JS

File: src/style/shared/dialog.less [hunk 3]
```diff
@@ -190,10 +190,20 @@ dialog::backdrop {
...
+    top: -0.9rem;
+    transform: rotate(45deg);
+}
+
+.pointerMixin (@atTop) when (default()) {
```
Comment (volfied) [inline src/style/shared/dialog.less]: This is all the cases not accounted for `.pointerMixin`

File: src/style/shared/dialog.less [hunk 4]
```diff
@@ -190,10 +190,20 @@ dialog::backdrop {
...
+    bottom: -0.9rem;
+    transform: rotate(-135deg);
+}
+
+.dialog-pointer(@left, @atTop) {
+    .pointerMixin(@atTop);
```
Comment (volfied) [inline src/style/shared/dialog.less]: And this is how you call a prepared mixin, adding the properties it calculates into the class.

**Comment by iwehrman** [comment]
Here's a failure case involving a short window:
image

**Comment by iwehrman** [comment]
Note that in the case above I think it should just show neither pointer.

**Comment by volfied** [comment]
@iwehrman thanks for the catch, I changed the class names to be more explicit and fixed the issue so it doesn't draw any pointers if it's going to float.

File: src/js/jsx/shared/Dialog.jsx [hunk 5]
```diff
@@ -358,8 +360,10 @@ define(function (require, exports, module) {
         render: function () {
             var children,
                 globalClass = (this.props.position === POSITION_METHODS.CENTER) ? "dialog__center" : "dialog__target",
+                // We reverse the pointerAbove/Below classes as they remove the dialog pointer rather than add it
                 classes = classnames(globalClass, this.props.className || "",
-                    { "placed-above": this.state.placedAbove }),
+                    { "pointer-above": !this.state.pointerBelow },
+                    { "pointer-below": !this.state.pointerAbove }),
```
Comment (iwehrman) [inline src/js/jsx/shared/Dialog.jsx]: Hmm, this is kind of confusing. Is it possible for these two sets of names to be more in sync?

Comment (volfied) [inline src/js/jsx/shared/Dialog.jsx]: Fixed it, thanks.

**Comment by volfied** [comment]
Back to you, @iwehrman

**Comment by iwehrman** [comment]
image

**Comment by volfied** [comment]
After a remote debugging session, we realized this issue happens when the swatch is half rendered outside the panel bounds (bottom one in the screenshot). Making a note here for myself.

**Comment by iwehrman** [comment]
A fix for this may also address #3440.

**Comment by volfied** [comment]
@iwehrman I can't reproduce this again, I've tried at all window sizes / color picker locations.

**Comment by iwehrman** [comment]
Hmm, I also can't repro at the moment. I guess that means it's fixed!

**Pull request merged**